c10/util/Half.h

```diff
@@ -355,6 +105,377 @@ inline uint16_t native_fp16_from_fp32_value(float f) {
   return fp16_to_bits(static_cast<float16_t>(f));
 }
 #endif
+//*******************************************************************************
+
+
+#if defined(X86_F16) && (defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512)) && !defined(__APPLE__)
+static inline uint16_t _f2h(float val) {
+#if defined(CPU_CAPABILITY_AVX2)
+#if defined(_MSC_VER)
+  __m256 v = _mm256_set1_ps(val);
+  __m128i o =
+      _mm256_cvtps_ph(v, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  return static_cast<std::uint16_t>(_mm_cvtsi128_si32(o));
+#else
+  return _cvtss_sh(val, _MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC);
+#endif
+#elif defined(CPU_CAPABILITY_AVX512)
+  __m512 v = _mm512_set1_ps(val);
+  __m256i o =
+      _mm512_cvtps_ph(v, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  return static_cast<std::uint16_t>(
+      _mm_cvtsi128_si32(_mm256_castsi256_si128(o)));
```

Comment:
Just curious, what is the benefit of having using Zmm, where Xmm is already an overkill? (I.e. AVX512 is true superset of AVX256, isn't it?) I.e. wouldn't it be faster to always use `_cvtss_sh` intrinsic?

Reply:
That's what I had originally in my PR. I think if you have an AVX 512, maybe the fact that you only touch a part of the register creates a problem that results in slower execution? I'll assume that Intel optimized the sequence end to end?

We had that problem on mainframe, and we ended up having separate rename engines for lower half and upper half of the newly extended 64b registers...

Because when an AVX 512 function calls an AVX2 function, the AVX2 function can't save the full register (it only is aware of the lower half), so the upper half of a callee-saved register would be destroyed.  (We actually spec'ed that to happen for some VSX registers to sidestep the rename issue on PPC, but you can only do that when you define a new API...)